packages/query-core/src/queryClient.ts

```diff
@@ -130,13 +131,22 @@ export class QueryClient {
     TData = TQueryFnData,
     TQueryKey extends QueryKey = QueryKey,
   >(
-    options: FetchQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+    options: EnsureQueryDataOptions<TQueryFnData, TError, TData, TQueryKey>,
   ): Promise<TData> {
     const cachedData = this.getQueryData<TData>(options.queryKey)
 
-    return cachedData !== undefined
-      ? Promise.resolve(cachedData)
-      : this.fetchQuery(options)
+    if (cachedData === undefined) return this.fetchQuery(options)
+    else {
+      const defaultedOptions = this.defaultQueryOptions(options)
+      const query = this.#queryCache.build(this, defaultedOptions)
+
+      if (query.isStaleByTime(defaultedOptions.staleTime) && options.revalidateIfStale) {
```

Comment:
let's switch the check for an early bailout of the if statement:

```suggestion
      if (options.revalidateIfStale && query.isStaleByTime(defaultedOptions.staleTime)) {
```